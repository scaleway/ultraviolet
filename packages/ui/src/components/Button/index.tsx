'use client'

import type { Theme } from '@emotion/react'
import styled from '@emotion/styled'
import { Icon } from '@ultraviolet/icons/legacy'
import type {
  AriaRole,
  ButtonHTMLAttributes,
  ComponentProps,
  MouseEventHandler,
  ReactNode,
  Ref,
} from 'react'
import { forwardRef } from 'react'
import type { ExtendedColor } from '../../theme'
import { Loader } from '../Loader'
import { Tooltip } from '../Tooltip'
import { SIZE_GAP_KEY, SIZE_HEIGHT, SIZE_PADDING_KEY } from './constants'

type ButtonSize = keyof typeof SIZE_HEIGHT

export const buttonSizes = Object.keys(SIZE_HEIGHT) as ButtonSize[]

// FOCUS RING
const FOCUS_RING_KEY = {
  danger: 'focusDanger',
  info: 'focusInfo',
  neutral: 'focusNeutral',
  primary: 'focusPrimary',
  // @note: no focusSecondary so far, it will be added later, so far focusPrimary sounds fine
  secondary: 'focusPrimary',
  success: 'focusSuccess',
  warning: 'focusWarning',
  white: 'focusNeutral',
  black: 'focusNeutral',
} as const

const isMonochrome = (sentiment: ExtendedColor) =>
  sentiment === 'white' || sentiment === 'black'

// VARIANTS
type StyledButtonProps = Required<
  Pick<
    FinalProps,
    'size' | 'sentiment' | 'disabled' | 'iconPosition' | 'fullWidth'
  >
> & { iconOnly: boolean }

const coreStyle = ({
  theme,
  size,
  sentiment,
  iconPosition,
  fullWidth,
  disabled,
  iconOnly,
}: { theme: Theme } & StyledButtonProps) => {
  const font =
    size === 'large'
      ? theme.typography.bodyStrong
      : theme.typography.bodySmallStrong

  let width = 'auto'
  if (fullWidth) {
    width = '100%'
  } else if (iconOnly) {
    width = `${theme.sizing[SIZE_HEIGHT[size]]}`
  }

  return `
    display: inline-flex;
    position: relative;
    height: ${theme.sizing[SIZE_HEIGHT[size]]};
    padding: 0 ${theme.space[SIZE_PADDING_KEY[size]]};
    flex-direction: ${iconPosition === 'right' ? 'row-reverse' : 'row'};
    gap: ${theme.space[SIZE_GAP_KEY[size]]};
    border-radius: ${theme.radii.default};
    box-sizing: border-box;
    width: ${width};
    align-items: center;
    cursor: ${disabled ? 'not-allowed' : 'pointer'};
    justify-content: center;
    outline-offset: 2px;
    white-space: nowrap;
    text-decoration: none;
    &:hover {
      text-decoration: none;
    }


    ${
      disabled
        ? ''
        : `
            &:active {
              box-shadow: ${theme.shadows[FOCUS_RING_KEY[sentiment]]};
            }
          `
    }

    /* We can't use Text component because of button hover effect, so we need to duplicate */
    font-size: ${font.fontSize};
    font-family: ${font.fontFamily};
    font-weight: ${font.weight};
    letter-spacing: ${font.letterSpacing};
    line-height: ${font.lineHeight};
    paragraph-spacing: ${font.paragraphSpacing};
    text-case: ${font.textCase};
  `
}

const StyledFilledButton = styled('button', {
  shouldForwardProp: prop =>
    !['size', 'sentiment', 'iconPosition', 'fullWidth', 'iconOnly'].includes(
      prop,
    ),
})<StyledButtonProps>`
  ${args => coreStyle(args)}

  background: ${({ theme, sentiment }) =>
    !isMonochrome(sentiment)
      ? theme.colors[sentiment].backgroundStrong
      : theme.colors.other.monochrome[sentiment].background};
  border: none;
  color: ${({ theme, sentiment }) => (!isMonochrome(sentiment) ? theme.colors[sentiment].textStrong : theme.colors.other.monochrome[sentiment === 'white' ? 'black' : 'white'].text)};

  ${({ theme, sentiment, disabled }) =>
    disabled
      ? `
            background: ${!isMonochrome(sentiment) ? theme.colors[sentiment].backgroundStrongDisabled : theme.colors.other.monochrome[sentiment].backgroundDisabled};
            color:
              ${!isMonochrome(sentiment) ? theme.colors[sentiment].textStrongDisabled : theme.colors.other.monochrome[sentiment].textDisabled};
        `
      : `
            &:hover, &:active
            {
                background: ${!isMonochrome(sentiment) ? theme.colors[sentiment].backgroundStrongHover : theme.colors.other.monochrome[sentiment].backgroundHover};
                color:
                ${!isMonochrome(sentiment) ? theme.colors[sentiment].textStrongHover : theme.colors.other.monochrome[sentiment === 'white' ? 'black' : 'white'].textHover};
            }
  `}
`

const StyledOutlinedButton = styled('button', {
  shouldForwardProp: prop =>
    !['size', 'sentiment', 'iconPosition', 'fullWidth', 'iconOnly'].includes(
      prop,
    ),
})<StyledButtonProps>`
  ${args => coreStyle(args)}

  background: none;
  border: 1px solid
    ${({ theme, sentiment }) =>
      !isMonochrome(sentiment)
        ? theme.colors[sentiment][
            sentiment === 'neutral' ? 'borderStrong' : 'border'
          ]
        : theme.colors.other.monochrome[sentiment].border};
  color: ${({ theme, sentiment }) => (!isMonochrome(sentiment) ? theme.colors[sentiment].text : theme.colors.other.monochrome[sentiment].text)};

  ${({ theme, sentiment, disabled }) =>
    disabled
      ? `
        color:
          ${!isMonochrome(sentiment) ? theme.colors[sentiment].textDisabled : theme.colors.other.monochrome[sentiment].textDisabled};
        border: 1px solid ${
          !isMonochrome(sentiment)
            ? theme.colors[sentiment][
                sentiment === 'neutral'
                  ? 'borderStrongDisabled'
                  : 'borderDisabled'
              ]
            : theme.colors.other.monochrome[sentiment].borderDisabled
        };

    `
      : `
        &:hover, &:active
       {
            background: ${!isMonochrome(sentiment) ? theme.colors[sentiment].backgroundHover : theme.colors.other.monochrome[sentiment].backgroundHover};
            color:
            ${!isMonochrome(sentiment) ? theme.colors[sentiment].textHover : theme.colors.other.monochrome[sentiment === 'white' ? 'black' : 'white'].textHover};
            border: 1px solid ${
              !isMonochrome(sentiment)
                ? theme.colors[sentiment][
                    sentiment === 'neutral'
                      ? 'borderStrongHover'
                      : 'borderHover'
                  ]
                : theme.colors.other.monochrome[sentiment].borderHover
            };

        }
`};
`

const StyledGhostButton = styled('button', {
  shouldForwardProp: prop =>
    !['size', 'sentiment', 'iconPosition', 'fullWidth', 'iconOnly'].includes(
      prop,
    ),
})<StyledButtonProps>`
  ${args => coreStyle(args)}

  background: none;
  border: none;
  color: ${({ theme, sentiment }) => (!isMonochrome(sentiment) ? theme.colors[sentiment].text : theme.colors.other.monochrome[sentiment].text)};

  ${({ theme, sentiment, disabled }) =>
    disabled
      ? `
        color:
          ${!isMonochrome(sentiment) ? theme.colors[sentiment].textDisabled : theme.colors.other.monochrome[sentiment].textDisabled};
      `
      : `
        &:hover, &:active
        {
            background: ${!isMonochrome(sentiment) ? theme.colors[sentiment].backgroundHover : theme.colors.other.monochrome[sentiment].backgroundHover};
            color:
              ${!isMonochrome(sentiment) ? theme.colors[sentiment].textHover : theme.colors.other.monochrome[sentiment === 'white' ? 'black' : 'white'].textHover};
        }
`}
`

const VARIANTS_COMPONENTS = {
  filled: {
    button: StyledFilledButton,
    link: StyledFilledButton.withComponent('a'),
  },
  outlined: {
    button: StyledOutlinedButton,
    link: StyledOutlinedButton.withComponent('a'),
  },
  ghost: {
    button: StyledGhostButton,
    link: StyledGhostButton.withComponent('a'),
  },
}

type ButtonVariant = keyof typeof VARIANTS_COMPONENTS
export const buttonVariants = Object.keys(
  VARIANTS_COMPONENTS,
) as ButtonVariant[]

type CommonProps = {
  type?: ButtonHTMLAttributes<HTMLButtonElement>['type']
  autoFocus?: ButtonHTMLAttributes<HTMLButtonElement>['autoFocus']
  variant?: ButtonVariant
  role?: AriaRole
  size?: ButtonSize
  className?: string
  'data-testid'?: string
  sentiment?: ExtendedColor
  disabled?: boolean
  /**
   * @deprecated `iconPosition` prop is deprecated and will be removed in the next major version.
   * You can directly set the icon into the children of the Button component.
   */
  iconPosition?: 'left' | 'right'
  /**
   * @deprecated `iconVariant` prop is deprecated and will be removed in the next major version.
   * You can directly set the icon into the children of the Button component.
   */
  iconVariant?: ComponentProps<typeof Icon>['variant']
  fullWidth?: boolean
  isLoading?: boolean
  'aria-label'?: string
  'aria-current'?: boolean
  'aria-controls'?: string
  'aria-expanded'?: boolean
  'aria-haspopup'?: boolean
  onClick?: MouseEventHandler<HTMLElement>
  tooltip?: string
  tabIndex?: ButtonHTMLAttributes<HTMLButtonElement>['tabIndex']
  onMouseDown?: MouseEventHandler<HTMLElement>
  onMouseUp?: MouseEventHandler<HTMLElement>
  onMouseOut?: MouseEventHandler<HTMLElement>
  onMouseEnter?: MouseEventHandler<HTMLElement>
  onMouseLeave?: MouseEventHandler<HTMLElement>
}

// @note: using XOR utility was generating some lint erros
type FinalProps = CommonProps &
  (
    | {
        // Button : Children + optional Icon
        children: ReactNode
        /**
         * @deprecated `iconVariant` prop is deprecated and will be removed in the next major version.
         * You can directly set the icon into the children of the Button component.
         */
        icon?: ComponentProps<typeof Icon>['name']
        name?: string
        href?: never
        target?: never
        download?: never
      }
    | {
        // Button : Icon only
        children?: never
        /**
         * @deprecated `iconVariant` prop is deprecated and will be removed in the next major version.
         * You can directly set the icon into the children of the Button component.
         */
        icon: ComponentProps<typeof Icon>['name']
        name?: string
        href?: never
        target?: never
        download?: never
      }
    | {
        // Anchor : Children + optional Icon
        children: ReactNode
        /**
         * @deprecated `iconVariant` prop is deprecated and will be removed in the next major version.
         * You can directly set the icon into the children of the Button component.
         */
        icon?: ComponentProps<typeof Icon>['name']
        name?: never
        href: string
        target?: string
        download?: string
      }
    | {
        // Anchor : Children + Icon Only
        children?: never
        /**
         * @deprecated `iconVariant` prop is deprecated and will be removed in the next major version.
         * You can directly set the icon into the children of the Button component.
         */
        icon: ComponentProps<typeof Icon>['name']
        name?: never
        href: string
        target?: string
        download?: string
      }
  )

/**
 * Button component is used to trigger an action or event, such as submitting a form, opening a dialog,
 * canceling an action, or performing a delete operation.
 */
export const Button = forwardRef<Element, FinalProps>(
  (
    {
      type = 'button',
      className,
      'data-testid': dataTestId,
      sentiment = 'primary',
      variant = 'filled',
      size = 'large',
      disabled = false,
      icon,
      iconPosition = 'left',
      iconVariant,
      fullWidth = false,
      isLoading = false,
      children,
      onClick,
      onMouseDown,
      onMouseUp,
      onMouseOut,
      onMouseEnter,
      onMouseLeave,
      name,
      'aria-label': ariaLabel,
      'aria-current': ariaCurrent,
      'aria-controls': ariaControls,
      'aria-expanded': ariaExpanded,
      'aria-haspopup': ariaHaspopup,
      href,
      download,
      target,
      role,
      tooltip,
      tabIndex,
      autoFocus,
    },
    ref,
  ) => {
    const computeIsDisabled = disabled || isLoading

    const content = (
      <>
        {!isLoading && icon ? (
          <Icon name={icon} size="small" variant={iconVariant} />
        ) : null}
        {isLoading ? (
          <Loader
            active
            trailColor="transparent"
            size="1em"
            color="currentColor"
          />
        ) : null}
        {children}
      </>
    )

    // @note: an anchor can't be disabled
    if (href && !computeIsDisabled) {
      const Component = VARIANTS_COMPONENTS[variant].link

      return (
        <Tooltip text={tooltip} containerFullWidth={fullWidth}>
          <Component
            role={role}
            className={className}
            data-testid={dataTestId}
            disabled={false}
            fullWidth={fullWidth}
            iconPosition={iconPosition}
            sentiment={sentiment}
            size={size}
            type={type}
            onClick={onClick}
            aria-label={ariaLabel}
            aria-current={ariaCurrent}
            aria-controls={ariaControls}
            aria-expanded={ariaExpanded}
            aria-haspopup={ariaHaspopup}
            href={href}
            target={target}
            download={download}
            ref={ref as Ref<HTMLAnchorElement>}
            iconOnly={!!icon && !children}
            tabIndex={tabIndex}
            autoFocus={autoFocus}
            onMouseDown={onMouseDown}
            onMouseUp={onMouseUp}
            onMouseOut={onMouseOut}
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
          >
            {content}
          </Component>
        </Tooltip>
      )
    }

    const Component = VARIANTS_COMPONENTS[variant].button

    return (
      <Tooltip text={tooltip} containerFullWidth={fullWidth}>
        <Component
          role={role}
          className={className}
          data-testid={dataTestId}
          disabled={computeIsDisabled}
          fullWidth={fullWidth}
          iconPosition={iconPosition}
          sentiment={sentiment}
          size={size}
          type={type}
          onClick={onClick}
          ref={ref as Ref<HTMLButtonElement>}
          name={name}
          aria-label={ariaLabel}
          aria-current={ariaCurrent}
          aria-controls={ariaControls}
          aria-expanded={ariaExpanded}
          aria-haspopup={ariaHaspopup}
          iconOnly={!!icon && !children}
          tabIndex={tabIndex}
          autoFocus={autoFocus}
          onMouseDown={onMouseDown}
          onMouseUp={onMouseUp}
          onMouseOut={onMouseOut}
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
        >
          {content}
        </Component>
      </Tooltip>
    )
  },
)
