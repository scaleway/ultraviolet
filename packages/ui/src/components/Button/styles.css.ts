import { theme } from '@ultraviolet/themes'
import type { RecipeVariants } from '@vanilla-extract/recipes'
import { recipe } from '@vanilla-extract/recipes'
import { SIZE_GAP_KEY, SIZE_HEIGHT, SIZE_PADDING_KEY } from './constants'

const { monochrome } = theme.colors.other

const sentiments = [
  'primary',
  'secondary',
  'danger',
  'info',
  'success',
  'warning',
  'neutral',
  'black',
  'white',
] as const

const sentimentThemeMap: Record<
  Exclude<(typeof sentiments)[number], 'black' | 'white'>,
  typeof theme.colors.primary
> = {
  primary: theme.colors.primary,
  secondary: theme.colors.secondary,
  danger: theme.colors.danger,
  info: theme.colors.info,
  success: theme.colors.success,
  warning: theme.colors.warning,
  neutral: theme.colors.neutral,
}

function getFilledStyle(sentiment: (typeof sentiments)[number]) {
  if (sentiment === 'black') {
    return {
      background: monochrome.black.background,
      color: monochrome.white.text,
      border: 'none',
      selectors: {
        '&:hover, &:active': {
          background: monochrome.black.backgroundHover,
          color: monochrome.white.textHover,
        },
        '&:disabled': {
          background: monochrome.black.backgroundDisabled,
          color: monochrome.black.textDisabled,
        },
      },
    }
  }
  if (sentiment === 'white') {
    return {
      background: monochrome.white.background,
      color: monochrome.black.text,
      border: 'none',
      selectors: {
        '&:hover, &:active': {
          background: monochrome.white.backgroundHover,
          color: monochrome.black.textHover,
        },
        '&:disabled': {
          background: monochrome.white.backgroundDisabled,
          color: monochrome.white.textDisabled,
        },
      },
    }
  }
  const selectedSentiment = sentimentThemeMap[sentiment]

  return {
    background: selectedSentiment.backgroundStrong,
    color: selectedSentiment.textStrong,
    border: 'none',
    selectors: {
      '&:hover, &:active': {
        background: selectedSentiment.backgroundStrongHover,
        color: selectedSentiment.textStrongHover,
      },
      '&:disabled': {
        background: selectedSentiment.backgroundStrongDisabled,
        color: selectedSentiment.textStrongDisabled,
      },
    },
  }
}

function getOutlinedStyle(sentiment: (typeof sentiments)[number]) {
  if (sentiment === 'black') {
    return {
      background: 'none',
      color: monochrome.black.text,
      border: `1px solid ${monochrome.black.border}`,
      selectors: {
        '&:hover, &:active': {
          background: monochrome.black.backgroundHover,
          color: monochrome.white.textHover,
          border: `1px solid ${monochrome.black.borderHover}`,
        },
        '&:disabled': {
          color: monochrome.black.textDisabled,
          border: `1px solid ${monochrome.black.borderDisabled}`,
        },
      },
    }
  }
  if (sentiment === 'white') {
    return {
      background: 'none',
      color: monochrome.white.text,
      border: `1px solid ${monochrome.white.border}`,
      selectors: {
        '&:hover, &:active': {
          background: monochrome.white.backgroundHover,
          color: monochrome.black.textHover,
          border: `1px solid ${monochrome.white.borderHover}`,
        },
        '&:disabled': {
          color: monochrome.white.textDisabled,
          border: `1px solid ${monochrome.white.borderDisabled}`,
        },
      },
    }
  }
  if (sentiment === 'neutral') {
    const selectedSentiment = sentimentThemeMap[sentiment]

    return {
      background: 'none',
      color: selectedSentiment.text,
      border: `1px solid ${selectedSentiment.borderStrong}`,
      selectors: {
        '&:hover, &:active': {
          background: selectedSentiment.backgroundHover,
          color: selectedSentiment.textHover,
          border: `1px solid ${selectedSentiment.borderStrongHover}`,
        },
        '&:disabled': {
          color: selectedSentiment.textDisabled,
          border: `1px solid ${selectedSentiment.borderStrongDisabled}`,
        },
      },
    }
  }
  const selectedSentiment = sentimentThemeMap[sentiment]

  return {
    background: 'none',
    color: selectedSentiment.text,
    border: `1px solid ${selectedSentiment.border}`,
    selectors: {
      '&:hover, &:active': {
        background: selectedSentiment.backgroundHover,
        color: selectedSentiment.textHover,
        border: `1px solid ${selectedSentiment.borderHover}`,
      },
      '&:disabled': {
        color: selectedSentiment.textDisabled,
        border: `1px solid ${selectedSentiment.borderDisabled}`,
      },
    },
  }
}

function getGhostStyle(sentiment: (typeof sentiments)[number]) {
  if (sentiment === 'black') {
    return {
      background: 'none',
      color: monochrome.black.text,
      border: 'none',
      selectors: {
        '&:hover, &:active': {
          background: monochrome.black.backgroundHover,
          color: monochrome.white.textHover,
        },
        '&:disabled': {
          color: monochrome.black.textDisabled,
        },
      },
    }
  }
  if (sentiment === 'white') {
    return {
      background: 'none',
      color: monochrome.white.text,
      border: 'none',
      selectors: {
        '&:hover, &:active': {
          background: monochrome.white.backgroundHover,
          color: monochrome.black.textHover,
        },
        '&:disabled': {
          color: monochrome.white.textDisabled,
        },
      },
    }
  }
  const selectedSentiment = sentimentThemeMap[sentiment]

  return {
    background: 'none',
    color: selectedSentiment.text,
    border: 'none',
    selectors: {
      '&:hover, &:active': {
        background: selectedSentiment.backgroundHover,
        color: selectedSentiment.textHover,
      },
      '&:disabled': {
        color: selectedSentiment.textDisabled,
      },
    },
  }
}

const sharedSizeStyled = {
  fontSize: theme.typography.bodyStrong.fontSize,
  fontFamily: theme.typography.bodyStrong.fontFamily,
  fontWeight: theme.typography.bodyStrong.weight,
  letterSpacing: theme.typography.bodyStrong.letterSpacing,
  lineHeight: theme.typography.bodyStrong.lineHeight,
}

export const button = recipe({
  base: {
    display: 'inline-flex',
    position: 'relative',
    flexDirection: 'row',
    boxSizing: 'border-box',
    alignItems: 'center',
    justifyContent: 'center',
    outlineOffset: '2px',
    whiteSpace: 'nowrap',
    textDecoration: 'none',
    borderRadius: theme.radii.default,
    selectors: {
      '&:hover': {
        textDecoration: 'none',
      },
    },
  },
  variants: {
    size: {
      large: {
        height: theme.sizing[SIZE_HEIGHT.large],
        paddingLeft: theme.space[SIZE_PADDING_KEY.large],
        paddingRight: theme.space[SIZE_PADDING_KEY.large],
        gap: theme.space[SIZE_GAP_KEY.large],
        ...sharedSizeStyled,
      },
      medium: {
        height: theme.sizing[SIZE_HEIGHT.medium],
        paddingLeft: theme.space[SIZE_PADDING_KEY.medium],
        paddingRight: theme.space[SIZE_PADDING_KEY.medium],
        gap: theme.space[SIZE_GAP_KEY.medium],
        ...sharedSizeStyled,
      },
      small: {
        height: theme.sizing[SIZE_HEIGHT.small],
        paddingLeft: theme.space[SIZE_PADDING_KEY.small],
        paddingRight: theme.space[SIZE_PADDING_KEY.small],
        gap: theme.space[SIZE_GAP_KEY.small],
        ...sharedSizeStyled,
      },
      xsmall: {
        height: theme.sizing[SIZE_HEIGHT.xsmall],
        paddingLeft: theme.space[SIZE_PADDING_KEY.xsmall],
        paddingRight: theme.space[SIZE_PADDING_KEY.xsmall],
        gap: theme.space[SIZE_GAP_KEY.xsmall],
        ...sharedSizeStyled,
      },
    },
    fullWidth: {
      true: {
        width: '100%',
      },
      false: {
        width: 'auto',
      },
    },
    sentiment: {
      primary: {},
      secondary: {},
      danger: {},
      info: {},
      success: {},
      warning: {},
      neutral: {},
      black: {},
      white: {},
    },
    variant: {
      filled: {},
      outlined: {},
      ghost: {},
    },
    disabled: {
      true: {
        cursor: 'not-allowed',
      },
      false: {
        cursor: 'pointer',
      },
    },
  },
  compoundVariants: [
    ...sentiments.map(sentiment => ({
      variants: { variant: 'filled' as const, sentiment },
      style: getFilledStyle(sentiment),
    })),
    ...sentiments.map(sentiment => ({
      variants: { variant: 'outlined' as const, sentiment },
      style: getOutlinedStyle(sentiment),
    })),
    ...sentiments.map(sentiment => ({
      variants: { variant: 'ghost' as const, sentiment },
      style: getGhostStyle(sentiment),
    })),
  ],
  defaultVariants: {
    size: 'large',
    fullWidth: false,
    disabled: false,
    variant: 'filled',
    sentiment: 'primary',
  },
})

export type ButtonVariants = NonNullable<RecipeVariants<typeof button>>
