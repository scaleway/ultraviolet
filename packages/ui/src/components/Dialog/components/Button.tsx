import type { ComponentProps, ReactNode } from 'react'
import { Button } from '../../Button'
import { useDialogContext } from '../Context'

type DialogButtonProps = {
  children: ReactNode
} & Pick<ComponentProps<typeof Button>, 'onClick' | 'disabled' | 'tooltip'>

export const DialogButton = ({
  children,
  onClick,
  disabled,
  tooltip,
}: DialogButtonProps) => {
  const context = useDialogContext()

  return (
    <Button
      sentiment={context.sentiment}
      onClick={onClick}
      disabled={disabled}
      tooltip={tooltip}
    >
      {children}
    </Button>
  )
}
