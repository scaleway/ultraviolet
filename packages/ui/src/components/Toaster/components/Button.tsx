'use client'

import styled from '@emotion/styled'
import type { ComponentProps } from 'react'
import { Button } from '../../Button'

const StyledButton = styled(Button)`
  color: ${({ theme }) => theme.colors.neutral.textStronger};
`

export const ToastButton = ({
  variant,
  ...props
}: ComponentProps<typeof Button>) => {
  if (variant === 'ghost') {
    return <StyledButton {...props} size="xsmall" variant={variant} />
  }

  return <Button {...props} size="xsmall" variant={variant} />
}
