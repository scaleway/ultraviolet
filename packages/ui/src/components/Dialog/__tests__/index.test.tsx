import { screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import {
  renderWithTheme,
  shouldMatchEmotionSnapshotWithPortal,
} from '@utils/test'
import { describe, expect, it, vi } from 'vitest'
import { Button } from '../../Button'
import { Dialog } from '..'

describe('dialog', () => {
  it('should renders correctly', () =>
    shouldMatchEmotionSnapshotWithPortal(
      <Dialog open sentiment="primary" title="Test">
        <Dialog.Stack>
          <Dialog.Text>text example</Dialog.Text>
          <Dialog.Buttons
            primaryButton={<Button sentiment="danger">Discard changes</Button>}
            secondaryButton={
              <Dialog.CancelButton onClick={() => {}}>
                Cancel
              </Dialog.CancelButton>
            }
          />
        </Dialog.Stack>
      </Dialog>,
    ))

  it('should handle disclosure & render prop', async () => {
    renderWithTheme(
      <Dialog
        disclosure={<Button>Open Dialog</Button>}
        sentiment="primary"
        title="Title Test"
      >
        {({ close }) => (
          <Dialog.Stack>
            <Dialog.Text>text example</Dialog.Text>
            <Dialog.Buttons
              primaryButton={
                <Button sentiment="danger">Discard changes</Button>
              }
              secondaryButton={
                <Dialog.CancelButton onClick={close}>
                  Cancel
                </Dialog.CancelButton>
              }
            />
          </Dialog.Stack>
        )}
      </Dialog>,
    )

    const disclosure = screen.getByText('Open Dialog')
    expect(screen.queryByText('Title Test')).not.toBeInTheDocument()
    await userEvent.click(disclosure)
    expect(screen.getByText('Title Test')).toBeInTheDocument()
    const cancelButton = screen.getByText('Cancel')
    await userEvent.click(cancelButton)
    expect(screen.queryByText('Title Test')).not.toBeInTheDocument()
  })

  it('[CancelButton] : should handle click', async () => {
    const onClick = vi.fn()

    renderWithTheme(
      <Dialog.CancelButton onClick={onClick}>Cancel</Dialog.CancelButton>,
    )

    const button = screen.getByText('Cancel')
    await userEvent.click(button)
    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
