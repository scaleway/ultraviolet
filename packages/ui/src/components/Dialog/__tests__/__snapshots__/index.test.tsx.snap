// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dialog > should renders correctly 1`] = `
<DocumentFragment>
  <div
    data-testid="testing"
  />
</DocumentFragment>
`;
